libraries/chain/block_log.cpp

```diff
@@ -599,28 +652,53 @@ namespace eosio { namespace chain {
       EOS_ASSERT( fc::is_directory(data_dir) && fc::is_regular_file(data_dir / "blocks.log"), block_log_not_found,
-                 "Block log not found in '${blocks_dir}'", ("blocks_dir", data_dir)          );
+                  "Block log not found in '${blocks_dir}'", ("blocks_dir", data_dir)          );
 
       std::fstream  block_stream;
       block_stream.open( (data_dir / "blocks.log").generic_string().c_str(), LOG_READ );
 
       uint32_t version = 0;
       block_stream.read( (char*)&version, sizeof(version) );
-      EOS_ASSERT( version > 0, block_log_exception, "Block log was not setup properly." );
-      EOS_ASSERT( version >= min_supported_version && version <= max_supported_version, block_log_unsupported_version,
-                 "Unsupported version of block log. Block log version is ${version} while code supports version(s) [${min},${max}]",
-                 ("version", version)("min", block_log::min_supported_version)("max", block_log::max_supported_version) );
+      EOS_ASSERT( version >= block_log::min_supported_version && version <= block_log::max_supported_version, block_log_unsupported_version,
+                  "Unsupported version of block log. Block log version is ${version} while code supports version(s) [${min},${max}]",
+                  ("version", version)("min", block_log::min_supported_version)("max", block_log::max_supported_version) );
 
       uint32_t first_block_num = 1;
       if (version != 1) {
          block_stream.read ( (char*)&first_block_num, sizeof(first_block_num) );
       }
 
-      genesis_state gs;
-      fc::raw::unpack(block_stream, gs);
-      return gs;
+      return lambda(block_stream, version, first_block_num);
+   }
+
+   fc::optional<genesis_state> block_log::extract_genesis_state( const fc::path& data_dir ) {
+      return detail::block_log_impl::extract_chain_context<genesis_state>(data_dir, [](std::fstream& block_stream, uint32_t version, uint32_t first_block_num ) -> fc::optional<genesis_state> {
+         if (contains_genesis_state(version, first_block_num)) {
+            genesis_state gs;
+            fc::raw::unpack(block_stream, gs);
+            return gs;
+         }
+
+         // current versions only have a genesis state if they start with block number 1
+         return fc::optional<genesis_state>();
+      });
+   }
+
+   chain_id_type block_log::extract_chain_id( const fc::path& data_dir ) {
+      return *(detail::block_log_impl::extract_chain_context<chain_id_type>(data_dir, [](std::fstream& block_stream, uint32_t version, uint32_t first_block_num ) -> fc::optional<chain_id_type> {
+         // supported versions either contain a genesis state, or else the chain id only
+         if (contains_genesis_state(version, first_block_num)) {
+            genesis_state gs;
+            fc::raw::unpack(block_stream, gs);
+            return gs.compute_chain_id();
+         }
+         chain_id_type chain_id;
```

Comment:
Assert that it `contains_chain_id` because someone could be calling `extract_chain_id` on a version 1 or 2 block_log and they would currently get back garbage data.